modules/biobank/php/specimendao.class.inc

```diff
@@ -0,0 +1,719 @@
+            foreach ($row as $value) {
+                $mA[$row['methodId']][$row['attributeId']]['name']       = $row['name'];
+                $mA[$row['methodId']][$row['attributeId']]['datatypeId'] = $row['datatypeId'];
+                $mA[$row['methodId']][$row['attributeId']]['refTableId'] = $row['refTableId'];
+                $mA[$row['methodId']][$row['attributeId']]['required']   = $row['required'];
+            }
+        }
+
+        return $mA;
+    }
+
+    /**
+     * This function receives a Specimen Object, converts it into a SQL format
+     * and inserts it into all relevant tables in the database.
+     *
+     * @param object $specimen The Specimen Object to be saved the
+     *                         database.
+     *
+     * @return void
+     */
+    public function saveInstance(Specimen $specimen) : void
+    {
+        // Converts Specimen Object into SQL format.
+        $specimenArray = $this->getArrayFromInstance($specimen);
+        $data = $this->getSQLFromArray($specimenArray);
+
+        // TODO: Look into implementing a transaction here
+        // $this->db->_PDO->beginTransaction();
+        // $this->db->_PDO->commit();
+        // $this->db->_PDO->rollBack();
+
+        // Inserts or Updates the biobank_specimen table with respective data.
+        $this->db->insertOnDuplicateUpdate(
+            'biobank_specimen',
+            $data['biobank_specimen']
+        );
+
+        $specimenId = $specimen->getId();
+        if (isset($specimenId)) {
+            // If update:
+            // XXX: Logic will likely eventually go here.
+        } else {
+        // If insert:
+            $specimenId = $this->db->getLastInsertId();
+            // set the respective data array index to the value of the new
+            // specimen ID
+            $parentSpecimenIds = $newSpecimen->getParentSpecimenIds();
+            foreach ($parentSpecimenIds as $id) {
+                $data['biobank_specimen_parent'][$id]
+                    = array(
+                       'SpecimenID'       => $specimenId,
+                       'ParentSpecimenID' => $id,
+                      );
+            }
+            $data['biobank_specimen_freezethaw']['SpecimenID']  = $specimenId;
+            $data['biobank_specimen_collection']['SpecimenID']  = $specimenId;
+            $data['biobank_specimen_preparation']['SpecimenID'] = $specimenId;
+            $data['biobank_specimen_analysis']['SpecimenID']    = $specimenId;
+        }
+
+        // Insert or update biobank_specimen_collection with respective data.
+        $this->db->unsafeInsertOnDuplicateUpdate(
+            'biobank_specimen_collection',
+            $data['biobank_specimen_collection']
+        );
+
+        // If protocol exists, insert or update biobank_specimen_preparation
+        // with respective data.
+        if (isset($data['biobank_specimen_preparation']['SpecimenProtocolID'])) {
+            $this->db->unsafeInsertOnDuplicateUpdate(
```

Comment:
why `unsafe`?

Reply:
because I'm inserting a json object, which has trouble with the normal insert.

Reply:
Cool thanks for explaining. I'll make an issue about this.

Reply:
I opened #4374 to address this. Can you let me know if it works with your JSON data?